README.md
# Waka Profile Stats

## Add your dev metric into your profile readme

<p align="center">
  <img src="assets/example_one.png" />
  <h3 align="center">✨ Wonderfull Readme Stats</h3>
</p>

<p align="center">
   <img src="https://img.shields.io/badge/language-javascript-yellow"/>
   <img alt="GitHub license" src="https://img.shields.io/github/license/barthachijuu/waka-profile-stats">
   <img alt="GitHub Repo stars" src="https://img.shields.io/github/stars/barthachijuu/waka-profile-stats">
   <img src="https://img.shields.io/github/forks/barthachijuu/waka-profile-stats"/>
</p>
<p align="center">
   Are you human 👨‍💻 or like me a vampire 🧛?
   <br/>
   How many projects are you working on at the same time?
   <br/>
   What are languages you code in?
   <br/>
   Let's check out in your profile readme!
</p>

<p align="center">
    <a href="https://github.com/anmol098/barthachijuu/waka-profile-stats/issues">Report Bug</a>
    ·
    <a href="https://github.com/anmol098/barthachijuu/waka-profile-stats/issues">Request Feature</a>
  </p>
## How to start

1. You need to update your markdown file(.md) with 2 comments. You can refer [here](#update-readme) for updating it.
2. You'll need a WakaTime API Key. You can get that from your WakaTime Account Settings
    - You can refer [here](#new-to-wakatime), if is you're first time with WakaTime
3. You'll need a GitHub API Token with `repo` and `user` scope from [here](https://github.com/settings/tokens) if you're running the action to get commit metrics
   > enabling the `repo` scope seems **DANGEROUS**<br/>
   > but this GitHub Action only accesses your commit timestamp and lines of code added or deleted in repository you contributed.
   - You can use this [example](#example) to work it out
4. You need to save the WakaTime API Key and the GitHub API Token in the repository secrets. You can find that in the Settings of your repository. Be sure to save those as the following.
    - WakaTime API Key as `WAKATIME_API_KEY=<your wakatime API Key>`
    - GitHub Personal Access Token as `GH_TOKEN=<your github access token>`
5. You can enable and disable feature options based on requirements.

This Action will run every SUNDAY at 00.00 IST

## Update Readme

Add a comment to your `README.md` like this:

```md
<!--START_SECTION:waka-->
<!--END_SECTION:waka-->
```
These lines will be our entry-points for the dev metrics.

## New to WakaTime

WakaTime gives you an idea of the time you really spent on coding. This helps you boost your productivity and competitive edge.

- Head over to <https://wakatime.com> and create an account.
- Get your WakaTime API Key from your [Account Settings in WakaTime](https://wakatime.com/settings/account).
- Install the [WakaTime plugin](https://wakatime.com/plugins) in your favourite editor / IDE.
- Paste in your API key to start the analysis.

### Profile Repository

You'll need to get a [GitHub Access Token](https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token) with a `repo` and `user` scope and save it in the Repo Secrets `GH_TOKEN = <Your GitHub Access Token>`

Here is Sample Workflow File for running it:
```yml
name: Barthachijuu Profile Stats

on:
  workflow_dispatch:
  schedule:
    # Runs at 00:00AM UTC Every SUNDAY
    - cron: '0 0 * * SUN'

jobs:
  update-readme:
    name: Update this repo's README
    runs-on: ubuntu-latest
    steps:
      - uses: barthachijuu/waka-profile-stats@main
        env:
          WAKATIME_API_KEY: ${{ secrets.WAKATIME_API_KEY }}
          GH_TOKEN: ${{ secrets.GH_TOKEN }}
```
- Now you can commit and wait for run automatically, but you can also trigger to run it to see the result now. Just go to the `Actions` in your repo and select your `Profile Readme Development Stats` workflow and click in `Run workflow`. Now wait for a minute or two and you will see your changes.

## Extras

1. If you want to add the other info to your stats, you can add multiple `FLAGS` in your workflow file by default all flags are enabled
>except the lines of code flag due to heavy operation performed

```yml
  - uses: barthachijuu/waka-profile-stats@main
        env:
          WAKATIME_API_KEY: ${{ secrets.WAKATIME_API_KEY }}
          GH_TOKEN: ${{ secrets.GH_TOKEN }}
          SHOW_OS: "False"
          SHOW_PROJECTS: "False"
```

### Options Available

---

`INPUT_COMMIT_MESSAGE`        flag can be to set message commit, default is "Updated Readme with Waka Stats"

`SHOW_UPDATED_DATE`        flag can be set to `True` to show updated date in end of paragraph

`SHOW_TOTAL_TIME`       flag can be set to `True` to show the total time you have coded

![Code Time](http://img.shields.io/badge/Code_Time-37%20hrs%2041%20mins-blue)

`SHOW_PROFILE`       flag can be set to `False` to hide the Profile views

![Profile Views](http://img.shields.io/badge/Profile_Views-6-red)

`SHOW_COMMIT`       flag can be set to `False` to hide the commit stats

📆 **Maybe i am a vampire 🧛**
```text
🌞 Morning   0 commits      ░░░░░░░░░░░░░░░░░░░░░░░░░   0%

🌆 Daytime   22 commits     ██████████░░░░░░░░░░░░░░   41.51%

🌉 Evening   31 commits     ██████████████░░░░░░░░░░   58.49%

🌕 Night     0 commits      ░░░░░░░░░░░░░░░░░░░░░░░░░   0%
```

`SHOW_WEEK`       flag can be set to `False` to hide the commits made on different days of week

📅 **I do my best effort on** Thursday
 ```text
Monday       8 commits      ▒░░░░░░░░░░░░░░░░░░░░░░   9.76%

Tuesday      5 commits      ▒░░░░░░░░░░░░░░░░░░░░░░░   6.1%

Wednesday    16 commits     ████░░░░░░░░░░░░░░░░░░░░   19.51%

Thursday     37 commits     ███████████░░░░░░░░░░░░░   45.12%

Friday       16 commits     ████░░░░░░░░░░░░░░░░░░░░   19.51%

Saturday     0 commits      ░░░░░░░░░░░░░░░░░░░░░░░░░   0%

Sunday       0 commits      ░░░░░░░░░░░░░░░░░░░░░░░░░   0%
```

`SHOW_LANGUAGE`       flag can be set to `False` to hide the Coding Language You use

🏷️ ***Languages***
```text
JavaScript               13 hrs 2 mins       ████████████████████████   96.24%

Markdown                 28 mins             ▒░░░░░░░░░░░░░░░░░░░░░░░░   3.54%

Bash                     1 min               ▒░░░░░░░░░░░░░░░░░░░░░░░░   0.21%

JSON                     0 secs              ▒░░░░░░░░░░░░░░░░░░░░░░░░   0.01%
```

`SHOW_EDITORS`  flag can be set to `False` to hide the list of code-editors used

🧰 ***Editors***
```text
VS Code                  13 hrs 32 mins      █████████████████████████   100%
```

`SHOW_OS`       flag can be set to `False` to hide the OS details

📀 ***Operating System***
```text
Linux                    13 hrs 32 mins      █████████████████████████   100%
```

`SHOW_PROJECTS` flag can be set to `False` to hide the Projects worked on

💻 ***Projects***
```text
waka-profile-stats       13 hrs 32 mins      █████████████████████████   100%
```

`SHOW_LANGUAGE_PER_REPO`  flag can be set to `False` to hide the Number of repository in different language and frameworks

***I Mostly Code in*** CSS
```text
CSS                      1 repo              ████░░░░░░░░░░░░░░░░░░░░   16.67%

JavaScript               1 repo              ████░░░░░░░░░░░░░░░░░░░░   16.67%

Java                     1 repo              ████░░░░░░░░░░░░░░░░░░░░   16.67%

SCSS                     1 repo              ████░░░░░░░░░░░░░░░░░░░░   16.67%
```

`SHORT_INFO`  flag can be set to `False` to hide the short fun fact info of user
>This section requires personal access token with user permission otherwise data shown will be incorrect here

 **🤓 My Personal GitHub Info**

```properties
🏆 38 Personal contributions in the last year

🛡️ 207 Total contributions when i start a github profile

💾 5.86 kB Used in Github's Storage

🚫 Not Opted to hire

📖 11 Public repos

🔐 3 Private repos

🔃 15 Followers

🔄 19 Following

```

## 🎁 Support: Donate

> This project is **free**, **open source** and I try to provide excellent **free support**. Why donate? I work on this and other project for several hours in my spare time between my job and my lovely family, and try to keep it up to date and working. **THANK YOU!**

[![](https://img.shields.io/badge/donate-paypal-005EA6.svg?logo=paypal)](https://www.paypal.me/barthachijuu) [![](https://img.shields.io/badge/donate-patreon-F87668.svg?logo=patreon)](https://www.patreon.com/barthachijuu)  [![](https://img.shields.io/badge/donate-ko--fi-29abe0.svg?logo=ko-fi)](https://ko-fi.com/barthachijuu)

## 👑 Sponsors

Support this project by becoming a sponsor. 🙏 Become a sponsor on [patreon](https://www.patreon.com/join/barthachijuu) or become top3 sponsor on [ko-fi](https://ko-fi.com/barthachijuu). Your logo will show up here with a link to your website.

## 👔 Backers

Thank you to all our backers! 🙏 Become a backer on [patreon](https://www.patreon.com/join/barthachijuu).

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
<table><tr><td align="center"><a href="https://barthachijuu.dev"><img src="https://avatars3.githubusercontent.com/u/1281894?v=4" width="100px;" alt="Bartolo Amico"/><br /><sub><b>Bartolo Amico</b></sub></a><br /><a href="https://github.com/barthachijuu/circlemenu/commits?author=barthachijuu" title="Code">💻</a> <a href="#content-barthachijuu" title="Content">🖋</a> <a href="https://github.com/barthachijuu/circlemenu/commits?author=barthachijuu" title="Documentation">📖</a> <a href="#ideas-barthachijuu" title="Ideas, Planning, & Feedback">🤔</a> <a href="#infra-barthachijuu" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a> <a href="#maintenance-barthachijuu" title="Maintenance">🚧</a> <a href="#plugin-barthachijuu" title="Plugin/utility libraries">🔌</a> <a href="#projectManagement-barthachijuu" title="Project Management">📆</a></td></tr></table>

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

## License

This project is licensed under the MIT license, Copyright (©) 2020 Bartolomeo Amico. For more information see `LICENSE.md`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.0] — 2022-01-10

### Release

Waka Profile Stats v1.0.0 is out.

#### Main

Very first version of the project. Probably need some improovemenst


query.js
const repoQuery = {
  userInfoQuery: `{
    viewer {
        login,
        id,
        email
      }
    }
`,
  list_repos: `{
  user(login: "$username") {
    repositories(orderBy: {field: CREATED_AT, direction: ASC}, last: 100, affiliations: [OWNER, COLLABORATOR, ORGANIZATION_MEMBER], isFork: false) {
      totalCount
      edges {
        node {
          object(expression:"master") {
                ... on Commit {
                history (author: { id: "$id" }){
                totalCount
                    }
                }
                }
          primaryLanguage {
            color
            name
            id
          }
          stargazers {
            totalCount
          }
          createdAt
          name
          owner {
            id
            login
          }
          nameWithOwner
        }
      }
    }
    location
    createdAt
    name
  }
}`,
contributedQuery: `{
    user(login: "$username") {
      repositoriesContributedTo(last: 100, includeUserRepositories: true) {
        nodes {
          isFork
          name
          owner {
            login
          }
        }
      }
    }
  }`,
  createCommitQuery: `query {
    repository(owner: "$owner", name: "$name") {
      defaultBranchRef {
        target {
          ... on Commit {
            history(first: 100, author: { id: "$id" }) {
              edges {
                node {
                  committedDate
                }
              }
            }
          }
        }
      }
    }
  }`
}

module.exports = repoQuery;


utils.js
const axios = require ('axios');

const defaultHeaders = {
    Accept: 'application/json, text/plain, */*',
    'cache-control': 'no-cache',
    'Content-Type': 'application/json',
    pragma: 'no-cache',
};

// Axios configuration
const axiosInstance = axios.create({
    baseURL: '',
    headers: {
        common: {
            ...defaultHeaders
        },
    },
});

const GITHUBURL = 'https://api.github.com';
const GRAPHQLURL = '/graphql';
const WAKAURL = 'https://wakatime.com/api/v1/users/current/';

// Add a request interceptor
axiosInstance.interceptors.request.use(config => config, error => Promise.reject(error));

// Add a response interceptor
axiosInstance.interceptors.response.use((response) => {
    // Do something with response data
    response.data.status = response.status;
    return response.data;
}, (error) => {
    return error.response;
});

const handlingResponse = (data, prop = 'response') => {
  const response = { error: '' };
  if (data && data.status !== 200) {
    switch (data.status) {
    case 401:
      data.type = 'Unauthorized';
      break;
    case 404:
      data.type = 'Not Found';
      break;
    case 500:
      data.type = 'ServerError';
      break;
    default:
      data.type = 'Invalid Data';
      break;
    }
    return { response, error: data.data };
  }
  delete response.error;
  response[prop] = 'data' in data ? data.data : data;
  return response;
};


const gitApi = async (url, token, prop, isCustomApi = false) => {
  if(isCustomApi){
    axiosInstance.defaults.baseURL = url;
    url = '';
  } else {
    axiosInstance.defaults.baseURL = GITHUBURL;
  }
  let data;
  try {
    data = await axiosInstance({
      headers: {
          Authorization: `token ${token}`,
      },
      url: url,
    });
    return handlingResponse(data, prop);
  } catch (error) {
    return handlingResponse(error);
  }
}

const gitApiGraphQl = async (token, query, vars = {}, prop) => {
  axiosInstance.defaults.baseURL = GITHUBURL;
  let data;

  try {
    data = await axiosInstance.post(GRAPHQLURL, {
      query,
      variables: vars
      }, {
        headers: {
          Authorization: `token ${token}`,
        },
      });
    return handlingResponse(data, prop);
  } catch (error) {
    return handlingResponse(error);
  }
}

const wakatimeApi = async (url, prop) => {
  axiosInstance.defaults.baseURL = WAKAURL;
  let data;
  try {
    data = await axiosInstance({
        url: url,
    });
    return handlingResponse(data, prop);
  } catch (error) {
      return handlingResponse(error);
  }
}

const substitute = (str, search, replace) => {
  let string = str;
  if (Array.isArray(search) && Array.isArray(replace)){
    search.forEach((s, k)  => {
      string = string.split(s).join(replace[k])
    });
  } else {
    string = str.split(search).join(replace);
  }

  return string;
};

const convertData = (value) => {
  const decimalSuffix = ['kB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
  const base = 1024;
  const bytes = parseFloat(value)
  const abs_bytes = Math.abs(bytes)

  if(value === 1) {
    return `${value} Byte`;
  } else if(value < base){
    return `${value} Bytes`;
  }

  let res = '';
  let find = false;
  decimalSuffix.forEach((s, k) => {
    const unit = base ** (k + 2);
    if(abs_bytes < unit && !find) {
      find = true;
      res = `${(base * bytes / unit).toFixed(2)} ${s}`
    }
  });
  return res;
};

exports.gitApi = gitApi;
exports.gitApiGraphQl = gitApiGraphQl;
exports.substitute = substitute;
exports.wakatimeApi = wakatimeApi;
exports.convertData = convertData;


index.js
#!/usr/bin/env node
require("dotenv").config();
const { Octokit } =  require('@octokit/rest');
const gitUtils = require('./utils');
const repoQuery = require('./query');

const {
  WAKATIME_API_KEY: wakatimeApiKey,
  GH_TOKEN: gitHubToken,
  SHOW_TOTAL_TIME: showTime,
  SHOW_PROFILE: showProfile,
  SHORT_INFO: shortInfo,
  SHOW_WAKASTAT: showWakaStat,
  SHOW_COMMIT: showCommit,
  SHOW_WEEK: showDayOfWeek,
  SHOW_LANGUAGE: showLanguage,
  SHOW_EDITORS: showEditors,
  SHOW_OS: showOs,
  SHOW_PROJECTS: showProject,
  SHOW_LANGUAGE_PER_REPO: showLanguagePerRepo,
  SHOW_UPDATE_DATE: showUpdateDate,
  INPUT_COMMIT_MESSAGE: commitMessage,
 } = process.env;

 const octokit = new Octokit({
    auth: gitHubToken,
 });

const temaplteTag='```';
let readme = '';
let stats = '';
const date = new Date;
let today, traffic, user, contrib, userStat, contributed;
const gitProfile = {
  userId: '',
  userName: '',
  userEmail: '',
}

const commitData = {
  owner: '',
  repo: '',
  message: commitMessage || 'Update Readme with Waka Stats'

}
START_COMMENT = '<!--START_SECTION:waka-->'
END_COMMENT = '<!--END_SECTION:waka-->'

function generateBarChart(perc) {
  const empty_block = "░";
  const done_block = "█";
  const mid_block = "▒"
  return `${(perc /4) < 4 && (perc /4) != 0 ? mid_block.repeat(1) : done_block.repeat((perc /4))}${empty_block.repeat((25 - (perc/4)))}`;
}

function makeStandardList (list){
  let string = `${temaplteTag}text`;
  for(const l of list){
    let lname = l.name.length;
    let ltext = l.text.length;
    string = `${string}\n${l.name}${(' ').repeat(25 - lname)}${l.text}${(' ').repeat(20 - ltext)}${generateBarChart(l.percent)}   ${Number(l.percent)}%\n`;
  }
  return `${string.substr(0, string.length - 1)}\n${temaplteTag}\n`;
}

function makeCommitList (list){
  let string = `${temaplteTag}text`;
  for(const l of list){
    let lname = l.name.length;
    let ltext = l.text.length;
    string = `${string}\n${l.name}${(' ').repeat(13 - lname)}${l.text}${(' ').repeat(15 - ltext)}${generateBarChart(l.percent)}   ${Number(l.percent)}%\n`;
  }
  return `${string.substr(0, string.length - 1)}\n${temaplteTag}\n`;
}

function initialize() {
  return new Promise((resolve, reject) => {
    gitUtils.gitApiGraphQl(gitHubToken, repoQuery.userInfoQuery).then(d => {
      gitProfile.userId = d.response.viewer.id;
      gitProfile.userEmail = d.response.viewer.email;
      gitProfile.userName = d.response.viewer.login;
      gitUtils.gitApi(`/repos/${gitProfile.userName}/${gitProfile.userName}/contents/README.md`, gitHubToken).then(d => {
        commitData.sha = d.response.sha;
        commitData.path = d.response.path;
        commitData.repo = gitProfile.userName;
        commitData.owner = gitProfile.userName;
        const buff = new Buffer.from(d.response.content, 'base64');
        const rdme = buff.toString('utf-8');
        resolve(rdme)
      });
    });
  });
}

function getRepos() {
  return new Promise((resolve, reject) => {
    gitUtils.gitApiGraphQl(gitHubToken, gitUtils.substitute(repoQuery.list_repos, ['$username', '$id'], [gitProfile.userName, gitProfile.userId]), {}, 'repos').then(d => {
      resolve(d.repos.user.repositories);
    });
  });
}

function getStats() {
  return new Promise((resolve, reject) => {
    let prom = [];
    let stats = '';
    if (showTime == 'true') {
      prom.push(gitUtils.wakatimeApi(`all_time_since_today?api_key=${wakatimeApiKey}`, 'today'));
    }
    if(showProfile == 'true') {
      prom.push(gitUtils.gitApi(`repos/${gitProfile.userName}/${gitProfile.userName}/traffic/views?per=week`, gitHubToken, 'traffic'));
    }
    if(shortInfo == 'true') {
      prom.push(gitUtils.gitApi(`user`, gitHubToken, 'user'));
      prom.push(gitUtils.gitApi(`https://github-contributions.now.sh/api/v1/${gitProfile.userName}`, gitHubToken, 'contrib', true));
    }
    if (showWakaStat == 'true') {
        prom.push(gitUtils.wakatimeApi(`stats/last_30_days?api_key=${wakatimeApiKey}`, 'userStat'));
      }
    if (showCommit == 'true') {
        prom.push(gitUtils.gitApiGraphQl(gitHubToken, gitUtils.substitute(repoQuery.contributedQuery, '$username', gitProfile.userName), {}, 'contributed'));
    }
    Promise.all(prom).then(values => {
      let statistics = '';
      let contribution = '';
      let traffics = '';
      let userInfo = '';
      let repos = [];
      [today, traffic, user, contrib, userStat, contributed] = values.map((el, key) => {
          return values[key][Object.keys(el)]
      })
      if(today) {
        statistics = `![Code Time](http://img.shields.io/badge/Code_Time-${today.text.replace(/\s/g, "%20")}-blue)   `
      }
      if(contrib){
        let contr = 0;
        if(contrib.years.length > 0) {
          contrib.years.forEach(c => {
            contr = contr + c.total;
          })
        }
        contribution = `🏆 ${contrib.years[0].total} Personal contributions in the last year\n\n🛡️ ${contr} Total contributions when i start a github profile\n`
      }
      if(traffic) {
        traffics = `![Profile Views](http://img.shields.io/badge/Profile_Views-${traffic.count}-red)\n\n`;
      }
      if(user) {
        userInfo = `💾 ${gitUtils.convertData(user.disk_usage)} Used in Github's Storage\n\n${user.hireable ? 'Not Opted to hire' : '🚫 Not Opted to hire'}\n\n📖 ${user.public_repos} Public repos \n\n🔐 ${user.total_private_repos !== null ? user.total_private_repos : 0} Private repos \n\n🔃 ${user.followers} Followers \n\n🔄 ${user.following} Following \n`;
      }
      stats = `${statistics}${traffics} **🤓 My Personal GitHub Info** \n\n${temaplteTag}properties\n${contribution.trimLeft()}\n${userInfo}\n${temaplteTag}`;
      if(contributed) {
        repos = contributed.user.repositoriesContributedTo.nodes.filter(a => !a.isFork);
      }
      const obj = {
        repos,
        stats,
        userStat
      }
       resolve(obj);
    }).catch (error => {
      console.log(error);
    });
  });
}

function generateCommitList(repos, stat) {
  return new Promise((resolve, reject) => {
    if(showCommit !== 'true') {
      resolve();
    }
    let string = ``;
    let sumAll = 0;
    let sum_week = 0;
    const weekDay = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];
    const dayPeriod = {
        morning: 0,
        daytime: 0,
        evening: 0,
        night: 0,
    }
    const dayOfWeek = {
        Sunday: 0,
        Monday: 0,
        Tuesday: 0,
        Wednesday: 0,
        Thursday: 0,
        Friday: 0,
        Saturday: 0,
    }
    const prom =[];
    for (let r of repos) {
      prom.push(gitUtils.gitApiGraphQl(gitHubToken, gitUtils.substitute(repoQuery.createCommitQuery, ['$owner', '$name', '$id'], [gitProfile.userName, r.name, gitProfile.userId])));
    }
    Promise.all(prom).then(values => {
      const no_activity = "No Activity Tracked This Week";
      let target = [];
      for (let c of values) {
        if (c.response.repository.defaultBranchRef != null) {
          target = target.concat(c.response.repository.defaultBranchRef.target.history.edges)
        }
      }
      for (let c of target) {
        const date = new Date(c.node.committedDate);
        const hour = date.getHours();
        const day = weekDay[date.getDay()];
        dayOfWeek[day] += 1;
        if (hour <= 0 && hour < 6) {
            dayPeriod.night += 1;
        } else if (hour <= 6 && hour < 12) {
            dayPeriod.morning += 1;
          } else if (hour <= 12 && hour < 18) {
            dayPeriod.daytime += 1;
        } else if (hour <= 18 && hour < 24) {
          dayPeriod.evening += 1;
        }
      }
      sumAll = Object.values(dayPeriod).reduce((a, b) => a + b);
      sum_week = Object.values(dayOfWeek).reduce((a, b) => a + b);
      const one_day = [{
          name: `🌞 Morning`,
          text: `${dayPeriod.morning} commits`,
          percent: ((dayPeriod.morning / sumAll) * 100).toFixed(2)
        },
        {
          name: `🌆 Daytime`,
          text: `${dayPeriod.daytime} commits`,
          percent: ((dayPeriod.daytime / sumAll) * 100).toFixed(2)
        },
        {
          name: `🌉 Evening`,
          text: `${dayPeriod.evening} commits`,
          percent: ((dayPeriod.evening / sumAll) * 100).toFixed(2)
        },
        {
          name: `🌕 Night`,
          text: `${dayPeriod.night} commits`,
          percent: ((dayPeriod.night / sumAll) * 100).toFixed(2)
        }];
      let dayTitle;
      dayTitle = dayPeriod.morning + dayPeriod.daytime > dayPeriod.evening + dayPeriod.night ? 'I am human 👨‍💻' : 'Maybe i am a vampire 🧛'
      string = `${string}\n📆 **${dayTitle}** \n${makeCommitList(one_day)}\n`;

      const day_of_week = [{
          name: `Monday`,
          text: `${dayOfWeek.Monday} commits`,
          percent: ((dayOfWeek.Monday / sum_week) * 100).toFixed(2)
        },
        {
          name: `Tuesday`,
          text: `${dayOfWeek.Tuesday} commits`,
          percent: ((dayOfWeek.Tuesday / sum_week) * 100).toFixed(2)
        },
        {
          name: `Wednesday`,
          text: `${dayOfWeek.Wednesday} commits`,
          percent: ((dayOfWeek.Wednesday / sum_week) * 100).toFixed(2)
        },
        {
          name: `Thursday`,
          text: `${dayOfWeek.Thursday} commits`,
          percent: ((dayOfWeek.Thursday / sum_week) * 100).toFixed(2)
        },
        {
          name: `Friday`,
          text: `${dayOfWeek.Friday} commits`,
          percent: ((dayOfWeek.Friday / sum_week) * 100).toFixed(2)
        },
        {
          name: `Saturday`,
          text: `${dayOfWeek.Saturday} commits`,
          percent: ((dayOfWeek.Saturday / sum_week) * 100).toFixed(2)
        },
        {
          name: `Sunday`,
          text: `${dayOfWeek.Sunday} commits`,
          percent: ((dayOfWeek.Sunday / sum_week) * 100).toFixed(2)
      }];
      if(showDayOfWeek == 'true'){
        let maxElement = {
          percent : 0,
        };
        let dayWeekTitle;
        for(let d of day_of_week) {
          if(Number(d.percent) > Number(maxElement.percent)) {
            maxElement = d;
          }
          dayWeekTitle = `**I do my best effort on** ${maxElement.name}`;
        }
        string = `${string}📅 ${dayWeekTitle}\n ${makeCommitList(day_of_week)}\n`;
      }
      if(showLanguage == 'true') {
        string = `${string}🏷️ ***Languages*** \n${stat.languages.length > 0 ? makeStandardList(stat.languages) : no_activity}\n`;
      }
      if(showEditors == 'true') {
        string = `${string}🧰 ***Editors*** \n${stat.editors.length > 0 ? makeStandardList(stat.editors) : no_activity}\n`;
      }
      if(showOs == 'true') {
        string = `${string}📀 ***Operating System*** \n${ stat.operating_systems.length > 0 ? makeStandardList(stat.operating_systems) : no_activity}\n`;
      }
      if(showProject == 'true') {
        const projects = stat.projects;
        projects.sort(function(a, b) {
            return b.percent - a.percent;
        });
        string = `${string}💻 ***Projects*** \n${ stat.projects.length > 0 ? makeStandardList(projects) : no_activity}\n`;
      }
      resolve(string)
    });
  });
}

function generateLanguagePerRepo(repos){
  return new Promise((resolve, reject) => {
    const lang = {};
    let total = 0;
    const data =[];
    for(let r of repos.edges) {
      if(r.node.primaryLanguage != null){
        const language = r.node.primaryLanguage.name;
        const color_code = r.node.primaryLanguage.color;
        total += 1;
        if(language in Object.keys(lang) === false) {
          lang[language] = {};
          lang[language]['count'] = 1
        } else {
          lang[language]['count'] += 1
        }
        lang[language]['color'] = color_code;
      }
    }
    const labels = Object.keys(lang).sort((a, b) => {return lang[b].count - lang[a].count});
    const mostLanguageRepo = labels[0];
    for(let label of labels){
      const perc = (lang[label].count / total * 100).toFixed(2);
      const extension = lang[label].count === 1 ? ' repo' : ' repos';
      data.push({ name: label, text: `${String(lang[label].count)}${extension}`, percent: perc });
    }
    const string = `***I Mostly Code in*** ${mostLanguageRepo} \n${makeStandardList(data)}\n`;
    resolve(string);
  });
}

function generateNewReadme(readme, stats) {
  return new Promise((resolve, reject) => {
    const regex = new RegExp(`(${START_COMMENT})[^]+(${END_COMMENT})[^*]`);
    const old = pat = readme.match(regex)[0];
    const string = readme.replace(old, `${START_COMMENT}\n${stats}\n${END_COMMENT}\n`);
    resolve(string);
  });
}


(async () => {
  let string = ``;
  console.log(`Start on ${date.toLocaleString()}`)
  readme = await initialize();
  const repos = await getRepos();
  stats = await getStats();
  const wakaStat = await generateCommitList(stats.repos, stats.userStat);
  string = `${string}${stats.stats}${wakaStat}`;
  if(showLanguagePerRepo == 'true') {
    string = `${string}${await generateLanguagePerRepo(repos)}\n\n`;
  }
  if(showUpdateDate == 'true'){
    const last_update = new Date();
    string = `${string}⌚ ***Last Stats Update on***\n${last_update.toUTCString()}`;
  }
  const newreadme = await generateNewReadme(readme, string);
  // commitData.content = new Buffer.from(newreadme).toString('base64');
  // const result = await octokit.repos.createOrUpdateFileContents(commitData);
  console.log(newreadme)
  console.log(`Readme updated ${result.status}`);
  const end_time = new Date;
  console.log(`End on ${end_time.toLocaleString()}`)
  console.log(`Program processed in ${Math.round((end_time.getTime() - date.getTime()) / 1000)} seconds\n`)
})();
